README.md
# SwiftyTimer

![Platforms](https://img.shields.io/badge/platforms-ios%20%7C%20osx%20%7C%20watchos%20%7C%20tvos-lightgrey.svg)
[![CI Status](https://api.travis-ci.org/radex/SwiftyTimer.svg?branch=master)](https://travis-ci.org/radex/SwiftyTimer)
[![CocoaPods](http://img.shields.io/cocoapods/v/SwiftyTimer.svg)](https://cocoapods.org/pods/SwiftyTimer)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](#carthage)
![Swift version](https://img.shields.io/badge/swift-4.2-orange.svg)

#### Modern Swifty API for `NSTimer`
###### SwiftyTimer allows you to instantly schedule delays and repeating timers using convenient closure syntax. It's time to get rid of Objective-C cruft.

Read [Swifty APIs: NSTimer](http://radex.io/swift/nstimer/) for more information about this project.

## Usage

You can easily schedule repeating and non-repeating timers (repeats and delays) using `Timer.every` and `Timer.after`:

```swift
Timer.every(0.7.seconds) {
    statusItem.blink()
}

Timer.after(1.minute) {
    println("Are you still here?")
}
```

You can specify time intervals with these intuitive helpers:

```swift
100.ms
1.second
2.5.seconds
5.seconds
10.minutes
1.hour
2.days
```

You can pass method references instead of closures:

```swift
Timer.every(30.seconds, align)
```

### Manual scheduling

If you want to make a timer object without scheduling, use `new(after:)` and `new(every:)`:

```swift
let timer = Timer.new(every: 1.second) {
    println(self.status)
}
```

(This should be defined as an initializer, but [a bug in Foundation](http://www.openradar.me/18720947) prevents this)

Call `start()` to schedule timers created using `new`. You can optionally pass the run loop and run loop modes:

```swift
timer.start()
timer.start(modes: .defaultRunLoopMode, .eventTrackingRunLoopMode)
```

### Invalidation

If you want to invalidate a repeating timer on some condition, you can take a `Timer` argument in the closure you pass in:

```swift
Timer.every(5.seconds) { (timer: Timer) in
    // do something
    
    if finished {
        timer.invalidate()
    }
}
```

## Installation

**Note:** If you're running Swift 2, use [SwiftyTimer v1.4.1](https://github.com/radex/SwiftyTimer/tree/1.4.1)

#### CocoaPods

If you're using CocoaPods, just add this line to your Podfile:

```ruby
pod 'SwiftyTimer'
```

Install by running this command in your terminal:

```sh
pod install
```

Then import the library in all files where you use it:

```swift
import SwiftyTimer
```

#### Carthage

Just add to your Cartfile:

```ruby
github "radex/SwiftyTimer"
```

#### Manually

Simply copy `Sources/SwiftyTimer.swift` to your Xcode project.

## More like this

If you like SwiftyTimer, check out [SwiftyUserDefaults](https://github.com/radex/SwiftyUserDefaults), which applies the same swifty approach to `NSUserDefaults`.

You might also be interested in my blog posts which explain the design process behind those libraries:
- [Swifty APIs: NSTimer](http://radex.io/swift/nstimer/)
- [Swifty APIs: NSUserDefaults](http://radex.io/swift/nsuserdefaults/)
- [Statically-typed NSUserDefaults](http://radex.io/swift/nsuserdefaults/static)
- [Swifty methods](http://radex.io/swift/methods/)

### Contributing

If you have comments, complaints or ideas for improvements, feel free to open an issue or a pull request.

### Author and license

Radek Pietruszewski

* [github.com/radex](http://github.com/radex)
* [twitter.com/radexp](http://twitter.com/radexp)
* [radex.io](http://radex.io)
* this.is@radex.io

SwiftyTimer is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
### 2.1.0 (2018-10-14)

This version supports Swift 4.2 without introducing changes to the library workflow.
Library source modifications include adopting new changes in the `RunLoopMode` enum introduced in Swift 4.2
- Updated for Swift 4.2 and Xcode 10 #45 @asowers1

### 2.0.0 (2016-09-23)

This is the Swift 3 update version.

It contains no major changes in the library itself, however it does change some APIs because of Swift 3 requirements.

- Updated for Swift 3 and Xcode 8 compatibility #28 @ldiqual

### 1.4.1 (2016-08-03)

- Add support for Xcode 8 (Swift 2.3) for Carthage users

### 1.4.0 (2016-04-10)

- Add a variant of `every` and `new(every:)` that takes a closure with `NSTimer` passed in
- Fix Carthage support for Mac (set deployment target to 10.9)

### 1.3.1 (2016-03-02)

- Added support for Swift Package Manager
- Refactoring (Removed NSTimerActor. Used CFRunLoopTimerCreateWithHandler instead.) #22 @Austinate
- Added Travis CI

### 1.3.0 (2016-02-29)

- Add Carthage support
- Add tvOS and watchOS support

### 1.2.0 (2015-09-18)

- Update to Swift 2
- Add millisecond helper (`100.ms`)

### 1.1.0 (2015-05-13)

- Add `start(runLoop:, modes:)`
- Refactoring

### 1.0.0 (2015-05-11)

- Initial release
- `NSTimer.after(...)` and `NSTimer.every(...)`
- `NSTimer.new`
- Ruby on Rails-inspired time helpers like (5.seconds or 1.minute)


Tests/SwiftyTimerTests/main.swift
import Cocoa

let app = NSApplication.sharedApplication()

class AppDelegate: NSObject, NSApplicationDelegate {
    func applicationDidFinishLaunching(aNotification: NSNotification) {
        test()
    }
    
    func test() {
        assert(1.second == 1.0)
        assert(1.minute == 60.0)
        assert(1.hour == 1.minute * 60)
        assert(1.2.seconds == 1.2)
        assert(1.5.minutes == 90.0)
        assert(1.5.hours == 5400.0)
        assert(1.3.milliseconds == 0.0013)
        assert(0.5.day == 43_200 )
        assert(1.day == 86_400 )
        assert(2.days == 172_800)
        test2()
    }
    
    func test2() {
        var fired = false
        NSTimer.after(0.1.seconds) {
            assert(!fired)
            fired = true
            self.test3()
        }
    }
    
    var timer1: NSTimer!
    
    func test3() {
        var fired = false
        timer1 = NSTimer.every(0.1.seconds) {
            if fired {
                self.test4()
                self.timer1.invalidate()
            } else {
                fired = true
            }
        }
    }
    
    let timer2 = NSTimer.new(after: 0.1.seconds) { fatalError() }
    let timer3 = NSTimer.new(every: 0.1.seconds) { fatalError() }
    
    func test4() {
        let timer = NSTimer.new(after: 0.1.seconds) {
            self.test5()
        }
        NSRunLoop.currentRunLoop().addTimer(timer, forMode: NSDefaultRunLoopMode)
    }
    
    var timer4: NSTimer!
    
    func test5() {
        var fired = false
        timer4 = NSTimer.new(every: 0.1.seconds) {
            if fired {
                self.timer4.invalidate()
                self.test6()
            } else {
                fired = true
            }
        }
        timer4.start()
    }
    
    func test6() {
        let timer = NSTimer.new(after: 0.1.seconds) {
            self.test7()
        }
        
        timer.start(runLoop: NSRunLoop.currentRunLoop(), modes: NSDefaultRunLoopMode, NSEventTrackingRunLoopMode)
    }

    func test7() {
        NSTimer.after(0.1.seconds, done)
    }
    
    func done() {
        print("All tests passed")
        app.terminate(self)
    }
}

let delegate = AppDelegate()
app.delegate = delegate
app.run()

Tests/SwiftyTimerTests/SwiftyTime.swift
//
// SwiftyTimer
//
// Copyright (c) 2015 Oktawian Chojnacki and Radosław Pietruszewski
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import Foundation

private let miliDivider: Double     = 1000
private let secondsInMinute: Double = 60
private let minutesInHour: Double   = 60
private let hoursInDay: Double      = 24

extension Double {
    
    private func assertOneOrLess(value: Double) -> Double {
        assert(self <= 1, "🤓 Use plural property for numbers above 1.")
        return value
    }
    
    public var milisecond:  NSTimeInterval { return assertOneOrLess(miliseconds) }
    public var miliseconds: NSTimeInterval { return self / miliDivider }
    public var ms:          NSTimeInterval { return miliseconds }
    
    public var second:      NSTimeInterval { return assertOneOrLess(seconds) }
    public var seconds:     NSTimeInterval { return self }
    
    public var minute:      NSTimeInterval { return assertOneOrLess(minutes) }
    public var minutes:     NSTimeInterval { return self * secondsInMinute }
    
    public var hour:        NSTimeInterval { return assertOneOrLess(hours)}
    public var hours:       NSTimeInterval { return minutes * minutesInHour }
    
    public var day:         NSTimeInterval { return assertOneOrLess(days) }
    public var days:        NSTimeInterval { return hours * hoursInDay }
}


Src/SwiftyTime.swift
//
// SwiftyTimer
//
// Copyright (c) 2015 Oktawian Chojnacki and Radosław Pietruszewski
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import Foundation

private let milliDivider: Double    = 1000
private let secondsInMinute: Double = 60
private let minutesInHour: Double   = 60
private let hoursInDay: Double      = 24

extension Double {
    
    private func assertOneOrLess(value: Double) -> Double {
        assert(self <= 1, "🤓 Use plural property for numbers above 1.")
        return value
    }
    
    public var millisecond: NSTimeInterval  { return assertOneOrLess(milliseconds) }
    public var milliseconds: NSTimeInterval { return self / milliDivider }
    public var ms: NSTimeInterval           { return milliseconds }
    
    public var second: NSTimeInterval       { return assertOneOrLess(seconds) }
    public var seconds: NSTimeInterval      { return self }
    
    public var minute: NSTimeInterval       { return assertOneOrLess(minutes) }
    public var minutes: NSTimeInterval      { return self * secondsInMinute }
    
    public var hour: NSTimeInterval         { return assertOneOrLess(hours)}
    public var hours: NSTimeInterval        { return minutes * minutesInHour }
    
    public var day: NSTimeInterval          { return assertOneOrLess(days) }
    public var days: NSTimeInterval         { return hours * hoursInDay }
}


Sources/SwiftyTimer.swift
//
// SwiftyTimer
//
// Copyright (c) 2015-2016 Radosław Pietruszewski
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import Foundation

extension Timer {
    
// MARK: Schedule timers
    
    /// Create and schedule a timer that will call `block` once after the specified time.
    
    @discardableResult
    public class func after(_ interval: TimeInterval, runLoop: RunLoop = .current, modes: RunLoopMode..., _ block: @escaping () -> Void) -> Timer {
        let timer = Timer.new(after: interval, block)
        timer.start(runLoop: runLoop, modes: modes)
        return timer
    }
    
    /// Create and schedule a timer that will call `block` repeatedly in specified time intervals.
    
    @discardableResult
    public class func every(_ interval: TimeInterval, runLoop: RunLoop = .current, modes: RunLoopMode..., _ block: @escaping () -> Void) -> Timer {
        let timer = Timer.new(every: interval, block)
        timer.start(runLoop: runLoop, modes: modes)
        return timer
    }
    
    /// Create and schedule a timer that will call `block` repeatedly in specified time intervals.
    /// (This variant also passes the timer instance to the block)
    
    @nonobjc @discardableResult
    public class func every(_ interval: TimeInterval, runLoop: RunLoop = .current, modes: RunLoopMode..., _ block: @escaping (Timer) -> Void) -> Timer {
        let timer = Timer.new(every: interval, block)
        timer.start(runLoop: runLoop, modes: modes)
        return timer
    }
    
// MARK: Create timers without scheduling
    
    /// Create a timer that will call `block` once after the specified time.
    ///
    /// - Note: The timer won't fire until it's scheduled on the run loop.
    ///         Use `NSTimer.after` to create and schedule a timer in one step.
    /// - Note: The `new` class function is a workaround for a crashing bug when using convenience initializers (rdar://18720947)

    public class func new(after interval: TimeInterval, _ block: @escaping () -> Void) -> Timer {
        return CFRunLoopTimerCreateWithHandler(kCFAllocatorDefault, CFAbsoluteTimeGetCurrent() + interval, 0, 0, 0) { _ in
            block()
        }
    }
    
    /// Create a timer that will call `block` repeatedly in specified time intervals.
    ///
    /// - Note: The timer won't fire until it's scheduled on the run loop.
    ///         Use `NSTimer.every` to create and schedule a timer in one step.
    /// - Note: The `new` class function is a workaround for a crashing bug when using convenience initializers (rdar://18720947)

    public class func new(every interval: TimeInterval, _ block: @escaping () -> Void) -> Timer {
        return CFRunLoopTimerCreateWithHandler(kCFAllocatorDefault, CFAbsoluteTimeGetCurrent() + interval, interval, 0, 0) { _ in
            block()
        }
    }
    
    /// Create a timer that will call `block` repeatedly in specified time intervals.
    /// (This variant also passes the timer instance to the block)
    ///
    /// - Note: The timer won't fire until it's scheduled on the run loop.
    ///         Use `NSTimer.every` to create and schedule a timer in one step.
    /// - Note: The `new` class function is a workaround for a crashing bug when using convenience initializers (rdar://18720947)
    
    @nonobjc public class func new(every interval: TimeInterval, _ block: @escaping (Timer) -> Void) -> Timer {
        var timer: Timer!
        timer = CFRunLoopTimerCreateWithHandler(kCFAllocatorDefault, CFAbsoluteTimeGetCurrent() + interval, interval, 0, 0) { _ in
            block(timer)
        }
        return timer
    }
    
// MARK: Manual scheduling
    
    /// Schedule this timer on the run loop
    ///
    /// By default, the timer is scheduled on the current run loop for the default mode.
    /// Specify `runLoop` or `modes` to override these defaults.
    
    public func start(runLoop: RunLoop = .current, modes: RunLoop.Mode...) {
        let modes = modes.isEmpty ? [.default] : modes
        
        for mode in modes {
            runLoop.add(self, forMode: mode)
        }
    }
    
    /// Schedule this timer on the run loop
    ///
    /// By default, the timer is scheduled on the current run loop for the default mode.
    /// Specify `runLoop` or `modes` to override these defaults.
    /// Usefull since the splat operator is not a thing in Swift
    
    public func start(runLoop: RunLoop = .current, modes: [RunLoopMode]) {
        let modes = modes.isEmpty ? [.defaultRunLoopMode] : modes
        
        for mode in modes {
            runLoop.add(self, forMode: mode)
        }
    }
}

// MARK: - Time extensions

extension Double {
    public var millisecond: TimeInterval  { return self / 1000 }
    public var milliseconds: TimeInterval { return self / 1000 }
    public var ms: TimeInterval           { return self / 1000 }
    
    public var second: TimeInterval       { return self }
    public var seconds: TimeInterval      { return self }
    
    public var minute: TimeInterval       { return self * 60 }
    public var minutes: TimeInterval      { return self * 60 }
    
    public var hour: TimeInterval         { return self * 3600 }
    public var hours: TimeInterval        { return self * 3600 }
    
    public var day: TimeInterval          { return self * 3600 * 24 }
    public var days: TimeInterval         { return self * 3600 * 24 }
}


SwiftyTimerTests/SwiftyTimerTests/main.swift
import Cocoa

let app = NSApplication.shared

class AppDelegate: NSObject, NSApplicationDelegate {
    func applicationDidFinishLaunching(_ aNotification: Notification) {
        test()
    }
    
    func test() {
        assert(1.second == 1.0)
        assert(1.minute == 60.0)
        assert(1.hour == 1.minute * 60)
        assert(1.2.seconds == 1.2)
        assert(1.5.minutes == 90.0)
        assert(1.5.hours == 5400.0)
        assert(1.3.milliseconds == 0.0013)
        assert(0.5.day == 43_200 )
        assert(1.day == 86_400 )
        assert(2.days == 172_800)
        test2()
    }
    
    func test2() {
        var fired = false
        Timer.after(0.1.seconds) {
            assert(!fired)
            fired = true
            self.test3()
        }
    }
    
    var timer1: Timer!
    
    func test3() {
        var fired = false
        timer1 = Timer.every(0.1.seconds) {
            if fired {
                self.test4()
                self.timer1.invalidate()
            } else {
                fired = true
            }
        }
    }
    
    let timer2 = Timer.new(after: 0.1.seconds) { fatalError() }
    let timer3 = Timer.new(every: 0.1.seconds) { fatalError() }
    
    func test4() {
        let timer = Timer.new(after: 0.1.seconds) {
            self.test5()
        }
        RunLoop.current.add(timer, forMode: .default)
    }
    
    var timer4: Timer!
    
    func test5() {
        var fired = false
        timer4 = Timer.new(every: 0.1.seconds) {
            if fired {
                self.timer4.invalidate()
                self.test6()
            } else {
                fired = true
            }
        }
        timer4.start()
    }
    
    func test6() {
        let timer = Timer.new(after: 0.1.seconds) {
            self.test7()
        }
        
        timer.start(runLoop: .current, modes: .default, .eventTracking)
    }

    func test7() {
        Timer.after(0.1.seconds, test8)
    }
    
    func test8() {
        var fires = 0
        let timer = Timer.new(every: 0.1.seconds) { (timer: Timer) in
            guard fires <= 1 else { fatalError("should be invalidated") }
            defer { fires += 1 }
            
            if fires == 1 {
                timer.invalidate()
                self.test9()
            }
        }
        timer.start()
    }
    
    func test9() {
        var fires = 0
        Timer.every(0.1.seconds) { (timer: Timer) in
            guard fires <= 1 else { fatalError("should be invalidated") }
            defer { fires += 1 }
            
            if fires == 1 {
                timer.invalidate()
                self.test10()
            }
        }
    }
    
    func test10() {
        var fires = 0
        Timer.every(0.1.seconds, modes: .commonModes) { (timer: Timer) in
            guard fires <= 1 else { fatalError("should be invalidated") }
            defer { fires += 1 }
            
            if fires == 1 {
                timer.invalidate()
                self.done()
            }
        }
    }
    
    func done() {
        print("All tests passed")
        app.terminate(self)
    }
}

let delegate = AppDelegate()
app.delegate = delegate
app.run()
